README.md
DESCRIPTION:
-----------
ASTRAL is a tool for estimating an unrooted species tree given a set of unrooted gene trees.
ASTRAL is statistically consistent under the multi-species coalescent model (and thus is useful for handling incomplete lineage sorting, i.e., ILS).
ASTRAL finds the species tree that has the maximum number of shared induced quartet trees with the set of gene trees, subject to the constraint that the set of bipartitions in the species tree comes from a predefined set of bipartitions. This predefined set is empirically decided by ASTRAL (but see tutorial on how to expand it). The current code corresponds to **ASTRAL-III** (see below for the publication).


The algorithm was designed by Tandy Warnow and Siavash Mirarab originally. ASTRAL-III incorporates many ideas by Chao Zhang and Maryam Rabiee.
[Code developers](https://github.com/smirarab/ASTRAL/graphs/contributors) are mainly Siavash Mirarab, Chao Zhang, Maryam Rabiee, and Erfan Sayyari.

Email: `astral-users@googlegroups.com` for questions.



## Publications:

#### Papers on the current version:
- Since version 5.1.1, the code corresponds to **ASTRAL-III**, described in:
    * Zhang, Chao, Maryam Rabiee, Erfan Sayyari, and Siavash Mirarab. 2018. “ASTRAL-III: Polynomial Time Species Tree Reconstruction from Partially Resolved Gene Trees.” BMC Bioinformatics 19 (S6): 153. [doi:10.1186/s12859-018-2129-y](https://doi.org/10.1186/s12859-018-2129-y).
- For **multi-individual** datasets, the relevant paper to cite is:
	* Rabiee, Maryam, Erfan Sayyari, and Siavash Mirarab. 2019. “Multi-Allele Species Reconstruction Using ASTRAL.” Molecular Phylogenetics and Evolution 130 (January). 286–96. [doi:10.1016/j.ympev.2018.10.033](https://doi.org/10.1016/j.ympev.2018.10.033).
- Since version 4.10.0, ASTRAL can also compute branch length (in coalescent units) and a measure of support called **local posterior probability**, described here:
    * Sayyari, Erfan, and Siavash Mirarab. 2016. “Fast Coalescent-Based Computation of Local Branch Support from Quartet Frequencies.” Molecular Biology and Evolution 33 (7): 1654–68.  [doi:10.1093/molbev/msw079](http://mbe.oxfordjournals.org/content/early/2016/05/12/molbev.msw079.short?rss=1)
- ASTRAL can also perform a **polytomy test** (`-t 10` option):
    * Sayyari, Erfan, and Siavash Mirarab. 2018. “Testing for Polytomies in Phylogenetic Species Trees Using Quartet Frequencies.” Genes 9 (3): 132. [doi:10.3390/genes9030132](http://www.mdpi.com/2073-4425/9/3/132).

- For practical tips on using ASTRAL see [this preprint book chapter](https://arxiv.org/pdf/1904.03826.pdf).

#### Papers on older versions:

- The original algorithm (ASTRAL-I) is described in:
    - Mirarab, Siavash, Rezwana Reaz, Md. Shamsuzzoha Bayzid, Théo Zimmermann, M. S. Swenson, and Tandy Warnow. 2014. “ASTRAL: Genome-Scale Coalescent-Based Species Tree Estimation.” Bioinformatics 30 (17): i541–48. [doi:10.1093/bioinformatics/btu462](doi.org/10.1093/bioinformatics/btu462).
- All the versions between 4.7.4  and 5.1.0 correspond to ASTRAL-II, described in:
    * Mirarab, Siavash, and Tandy Warnow. 2015. “ASTRAL-II: Coalescent-Based Species Tree Estimation with Many Hundreds of Taxa and Thousands of Genes.” Bioinformatics 31 (12): i44–52. [doi:10.1093/bioinformatics/btv234](http://bioinformatics.oxfordjournals.org/content/31/12/i44)


#### Papers with relevance to ASTRAL:
    
These papers do not describe features in ASTRAL, but are also relveant and we encourage you to read them:

1. **DiscoVista**: This paper shows how quartet scores (more broadly, genome discordance) can be visualized in interpretable ways. The visualization of quartet scores, in particular, is closely tied to the ASTRAL method. 
    - Sayyari, Erfan, J.B. James B. Whitfield, and Siavash Mirarab. 2018. “DiscoVista: Interpretable Visualizations of Gene Tree Discordance.” Molecular Phylogenetics and Evolution 122 (May): 110–15. [doi:10.1016/j.ympev.2018.01.019](https://doi.org/10.1016/j.ympev.2018.01.019).
- **Fragmentary data**: The following paper made the case that before inferring gene trees, removing fragmentary data (e.g., those that have uncharacteristically large numbers of gaps) should be removed. It also showed RAxML gene trees are preferable to FastTree trees. 
    - Sayyari, Erfan, James B Whitfield, and Siavash Mirarab. 2017. “Fragmentary Gene Sequences Negatively Impact Gene Tree and Species Tree Reconstruction.” Molecular Biology and Evolution 34 (12): 3279–91. [doi:10.1093/molbev/msx261](https://doi.org/10.1093/molbev/msx261).
- **Missing data**: The following paper showed that excluding genes because they have missing data is often detrimental to accuracy. 
	- Molloy, Erin K., and Tandy Warnow. 2018. “To Include or Not to Include: The Impact of Gene Filtering on Species Tree Estimation Methods.” Systematic Biology 67 (2): 285–303. [doi:10.1093/sysbio/syx077](https://doi.org/10.1093/sysbio/syx077).
- **TreeShrink**: This paper introduced a method for removing very long branches from gene trees in a statistically motivated way. These branches make gene trees less accurate. 
	-  Mai, Uyen, and Siavash Mirarab. 2018. “TreeShrink: Fast and Accurate Detection of Outlier Long Branches in Collections of Phylogenetic Trees.” BMC Genomics 19 (S5): 272. [doi:10.1186/s12864-018-4620-2](https://doi.org/10.1186/s12864-018-4620-2).
- **Sample Complexity**: This paper established the theoretical sample complexity (i.e., number of required genes) for ASTRAL.
    - Shekhar, Shubhanshu, Sebastien Roch, and Siavash Mirarab. 2018. “Species Tree Estimation Using ASTRAL: How Many Genes Are Enough?” IEEE/ACM Transactions on Computational Biology and Bioinformatics 15 (5): 1738–47. [doi:10.1109/TCBB.2017.2757930](https://doi.org/10.1109/TCBB.2017.2757930).
- **INSTRAL**: introduces an ASTRAL-based algorithm for adding new species unto an existing species tree; so, the phylogenetic placement problem but for species trees. 
	- Rabiee, Maryam, and Siavash Mirarab. 2018. “INSTRAL: Discordance-Aware Phylogenetic Placement Using Quartet Scores.” BioRxiv 432906. [doi:10.1101/432906](https://doi.org/10.1101/432906).
- **BestML:** This paper was published before ASTRAL but showed that using best ML gene trees is often preferable to using the consensus of running summary methods on bootstrapped gene trees. 
	- Mirarab, Siavash, Md Shamsuzzoha Bayzid, and Tandy Warnow. 2016. “Evaluating Summary Methods for Multilocus Species Tree Estimation in the Presence of Incomplete Lineage Sorting.” Systematic Biology 65 (3). Oxford University Press: 366–80. [doi:10.1093/sysbio/syu063](https://doi.org/10.1093/sysbio/syu063).


Documentations
-----------

- The rest of this README file
- **Our [tutorial](astral-tutorial.md)**.
- For practical tips on using ASTRAL, including on how to prepare input and interpret output, see [this paper](https://arxiv.org/pdf/1904.03826.pdf).
- The chapter of Siavash Mirarab's dissertation that describes ASTRAL in detail is provided [here](thesis-astral.pdf).
- Publications shown above have scientific details
- A [developer guide](developer-guide.md).

INSTALLATION:
-----------
* There is no installation required to run ASTRAL.
* Download using one of two approaches:
    * You simply need to download the [zip file](https://github.com/smirarab/ASTRAL/raw/master/Astral.5.6.7.zip) and extract the contents to a folder of your choice. 
    * Alternatively, you can clone the [github repository](https://github.com/smirarab/ASTRAL/). You then run `make.sh` to build the project or simply uncompress the zip file that is included with the repository.
* ASTRAL is a java-based application, and should run in any environment (Windows, Linux, Mac, etc.) as long as java is installed. 
  Java 1.5 or later is required. We have tested ASTRAL only on Linux and MAC.
* To test your installation, go to the place where you put the uncompressed ASTRAL, and run:

  ``` bash
   java -jar astral.5.6.7.jar -i test_data/song_primates.424.gene.tre
   ```

  This should quickly finish. There are also other sample input files under `test_data/` that can be used.

* ASTRAL can be run from any directory (e.g., `/path/to/astral/`). Then, you just need to run:

  ``` bash
  java -jar /path/to/astral/astral.5.6.7.jar
  ```

* Also, you can move `astral.5.6.7.jar` to any location you like and run it from there, but note that you need to move the `lib` directory with it as well.


EXECUTION:
-----------
ASTRAL currently has no GUI. You need to run it through the command-line. In a terminal, go the location where you have downloaded the software, and issue the following command:

```
  java -jar astral.5.6.7.jar
```

This will give you a list of options available in ASTRAL.

To find the species tree given a set of gene trees in a file called `in.tree`, use:

```
java -jar astral.5.6.7.jar -i in.tree
```

The results will be outputted to the standard output. To save the results in a file use the `-o` option (**Strongly recommended**):

```
java -jar astral.5.6.7.jar -i in.tree -o out.tre
```
To save the logs (**also recommended**), run:

```
java -jar astral.5.6.7.jar -i in.tree -o out.tre 2>out.log
```

###### Input: 
* The input gene trees are in the Newick format
* The input trees can have missing taxa, polytomies (unresolved branches), and also multiple individuals per species.
*  Taxon names cannot have quotation marks in their names (sorry!). This means you also cannot have weird characters like ? in the name (underscore is fine).
*  For performing a constrained search, you can provide the constraint tree in newick format to ASTRAL using the `-j` option.
* When multiple individuals from the same species are available, you can ask ASTRAL to force them to be together in the species tree. To do this, a mapping file needs to be provided using the `-a` option. This mapping file should have one line per species, and each line needs to be in one of two formats:

```
species_name [number of individuals] individual_1 individual_2 ...

species_name:individual_1,individual_2,...
```
   Note that when multiple individuals exist for the same species, your species name should be different from the individual names.
   
###### Output: 
The output in is Newick format and gives: 

* the species tree topology, 
* branch lengths in coalescent units (only for internal branches or for terminal branches if that species has multiple individuals),
* branch supports measured as [local posterior probabilities](http://mbe.oxfordjournals.org/content/early/2016/05/12/molbev.msw079.short?rss=1). 
* It can also annotate branches with other quantities, such as quartet support, as described in the [tutorial](astral-tutorial.md).

The ASTRAL tree leaves the branch length of terminal branches empty. Some tools for visualization and tree editing do not like this (e.g., ape). In FigTree, if you open the tree several times, it eventually opens up (at least on our machines). In ape, if you ask it to ignore branch lengths all together, it works. In general, if you tool does not like the lack of terminal branches, you can add a dummy branch length, [as in this script](https://github.com/smirarab/global/blob/master/src/mirphyl/utils/add-bl.py). 

### Other features (local posterior, bootstrapping):
Please refer to the [tutorial](astral-tutorial.md) for all other features, including bootstrapping, branch annotation, and local posterior probability.

### Memory:
For big datasets (say more than 5000 taxa), increasing the memory available to Java can result in speedups. Note that you should give Java only as much free memory as you have available on your machine. So, for example, if you have 3GB of free memory, you can invoke ASTRAL using the following command to make all the 3GB available to Java:

```
java -Xmx3000M -jar astral.5.6.7.jar -i in.tree
```

Acknowledgment
-----------
ASTRAL code uses bytecode and some reverse engineered code from PhyloNet package (with permission from the authors).


Bug Reports:
-----------
contact ``astral-users@googlegroups.com``


CHANGELOG.md
- version 5.7.0:
	- 	
- version 5.6.3:
	- add a zero length branch to the root branch that is not the outgroup. Needed for FigTree visualizaiton of multiind
	
- version 5.6.2:
	- added new option to remove bipartitions of a provided extra tree from search space

- version 5.6.1:
	- added a new option '-w x' to count each gene tree as 1/x when computing support values

- version 5.5.12: 
	- A spurious warning removed
	  
- version 5.5.11: 
     - An overflow error for trees with 10,000 leaves is fixed using an approach different from 5.5.10. 

- version 5.5.10: 
     - An overflow error for trees with 10,000 leaves is fixed. 

- version 5.5.9:
	- `-k` option was not working fixed	 
	- Yet another problem with branch lengths fixed: root branch length was sometimes not shown
	- A little bit of code refactoring performed
	 
- version 5.5.8:
	- Fixed a bug in the computation of branch lengths. The bug was introduced in version 5.5.7
	
- version 5.5.7:
    - Terminal branches for multi ind
    - Fixed a bug where for multi-ind datasets, the final quartet score was slightly underestimated. 
    - Remove more unneeded logs
     
- version 5.5.6:
   - Merge in DiscoVista changes
   - Update documentations and some help messages

- version 5.5.5:
   - Remove false alarms in the log file
    
- version 5.5.4:
   - Merge changes from multiind (Maryam) to fix a small bug and add polytomies limit size option

- version 5.5.3:
   - Fixed bug which affects efficiency  

- version 5.5.2:
   - Fixed memory consumption issue  

- version 5.5.1:
   - Combining version 5.5.0 and 4.11.2  

- version 5.5.0:
   - Combining version 5.4.4 and 5.3.*

- version 5.4.*: handling multi-individual cases

- version 5.3.*: 
  - further development on A* to form a 3-stage algorithm
  - faster annotation for polytomies
  
- version 5.2.5: ASTRAL-III - first correct version (conference proceedings)
  - Enlarge set X for datasets with polytomies; multiple samples; randomize greedy and UPGMA
  - For polytomies, do not half count  unresolved quartets
  - solved a bug where in rare cases, output polytomy could be unresolved due to ASTRAL-III changes

- version 5.2.4:
  - Bug fix: weight calculation

- version 5.2.3: development

- version 5.2.2: development

- version 5.2.1: development 

- version 5.2.0: ASTRAL-III - first attempt
    - polytree
    - better handling of polytomies
    - A*

- version 5.1.0:
     - Merge polytree and multiind

- version 5.0.3:
	- Bug fix: 5.0.2 has created a bug, which is fixed
	- Error messages changed to emphasize that quartet score for multi individuals needs adjustment and the first value printed is just for internal use
	- `-q` option now removes branch lengths that it does not compute (e.g., terminal ones and those close to polytomies)
	 
- version 5.0.2: 
	- Bug fix: root branch was not scored if it had degree two
	
- version 5.0.1: 
	- Improved documentation of the multi-ind branch
	- Some code refactoring impacting greedy consensus. May have changed speed a bit.
	 
- version 5.0.0:
    - Merge the master branch (4.10.12) and the multi-ind branch (4.8.0)
      This version should have all the multi-ind features of 4.8.0 and
      local posterior probability features of 4.10.12 (plus all other misc changes on the master branch)

- version 4.11.2:
   - Some small code refactoring in computing scores
   - Change the hidden (not published) option `-t 10` that performs polytomy tests.

- version 4.11.1:
   - Change the calculation of the similarity matrix to count unresolved quartets as half

- version 4.10.13:
   - A bug fixed where for input trees with polytomies, the distance matrix was not computed correctly. 

- version 4.10.12:
    - Update code documentation and some refactoring
    
- version 4.10.11:
    - A bug fixed where number of taxa = 64*X was throwing an error
    
- version 4.10.10:
	- Very significant improvement in the running time for datasets with many polytomies in input gene trees
	- Some running time improvement for datasets without polytomies 
	- A little bit of simplifying code refactoring
	 
- version 4.10.9:
	- A bug introduced in 4.10.8 is fixed. Version 4.10.8 simply did not run
	- Added `--gene-only` for gene-only bootstrapping
	 
- version 4.10.8:
	- Output taxon occupancy
	- Don't error out on gene trees with dummy degree 2 nodes
	- Check the species name against gene names for identity. 
	 
- version 4.10.7:
	- Small bug fix: with `-a` scoring using `-q` didn't work. Also, output both gene tree labels and the species tree labels. 
	- Bug fix: effective N and quartet scores were not correctly printed out with -t X options if gene trees had polytomies. posterior values were correctly computed.
	- Warning for low `EN` values.
	- Some spell checking on prompts
	
- version 4.10.6:
    - Small bug fix: with `-a` scoring didn't work
    - Print tree topology to stderr before scoring it. 
    - Warn for multi-individual datasets that this is not the right branch
    - Update tutorial

- version 4.10.5:
	- Hide `-y` option for now. Not ready for prime time yet. 
	
- version 4.10.4:
	- New option `-y` to score all possible trees; when used with `-t 12` outputs only the tree(s) with the best sum log local posterior
	- Change handling the scoring of root nodes
	- Add a bunch of for-test-only `-t` options
	 
- version 4.10.3:
	- *Important change:* change calculation of local PP for trees with missing data. Avoid up-weighting genes with fewer missing leaves. 
	- New feature: `-t 10` tests for polytomies. 
	
- version 4.10.2:
	- Bug fix: when in rome do as the americans do. For people using European machines, the support values were printed with a comma, wreaking havoc in the newick format. Has to use US format for those to fix this.


- version 4.10.1:
	- BUG FIX: Branch lengths could be negative in rare situations (few genes). This was because checks were for ML instead of MAP. 
	- Some extra stuff were outputted to the log. Removed. 

main/phylonet/coalescent/TreeCompletion.java
package phylonet.coalescent;

import java.io.IOException;
import java.io.StringReader;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.Stack;

import com.sun.org.apache.bcel.internal.generic.NEW;

import phylonet.lca.SchieberVishkinLCA;
import phylonet.tree.io.NewickReader;
import phylonet.tree.io.ParseException;
import phylonet.tree.model.TMutableNode;
import phylonet.tree.model.TNode;
import phylonet.tree.model.Tree;
import phylonet.tree.model.sti.STINode;
import phylonet.tree.model.sti.STITree;
import phylonet.tree.util.Trees;

class NodeInfo{
	private String color="";	
	private int greenDescendents = 0;
	public NodeInfo(String color, int greenDes){
		this.color = color;
		this.greenDescendents = greenDes;
	}
	public NodeInfo(int greenDes){
		this.greenDescendents = greenDes;
	}
	
	public String getColor() {
		return color;
	}
	public int getGrDes() {
		return greenDescendents;
	}
	
}
public class TreeCompletion {
	
	public static void main(String[] args) throws IOException, ParseException{
		//String tr1 = "((1,(2,3)a)b,(4,(5,(6,7)d)c));";
//		String tr1 = "((1,2)a,(3,(4,(6,(5,7))d)c));";
		//((3,99),((11,12,(22,1),30),(4,((5,9,18)a,6)))b,(7,2))c;

//		String tr1 = "((7,2)a,(3,(4,(6,(5,(9,(18,23,90))))d)c)f,(11,12,(22,1),30)g,99)r;";
//		String tr2 = "((3,99),((11,22,1),(5,9,18)a)b)c;";
		
//		String tr2 = "(193,(15,(62,(((99,43),(26,(106,(137,78)))),(((66,(3,(102,(151,165)))),(16,((186,((76,(124,(51,79))),(98,(123,((107,195),(131,(168,104))))))),(((95,121),(9,136)),((6,(67,41)),(81,((45,130),(198,(12,47))))))))),((((153,53),(159,44)),((183,42),((196,22),((19,(40,(14,110))),(156,((173,111),(13,185))))))),((83,((74,148),(146,(109,118)))),(((8,(50,10)),((24,88),((92,38),(132,(75,154))))),((31,((134,(157,82)),((49,72),((188,149),(119,(184,143)))))),(128,(((69,(((29,158),(71,96)),((179,28),((100,87),((126,122),(59,(120,23))))))),((181,((77,30),(152,(70,(85,46))))),((150,(187,(37,61))),((89,182),(161,(135,7)))))),(0,((((113,36),(63,34)),((172,178),((39,65),(68,(166,(60,86)))))),(((145,48),((127,(169,(171,(199,176)))),((32,(18,191)),(180,((142,(105,(117,175))),(((17,93),((162,52),(108,54))),((20,94),((138,(192,160)),(197,(200,141)))))))))),((((58,(114,167)),(57,(139,25))),(2,(((33,1),(190,21)),(5,((80,163),(55,(101,(189,84)))))))),(((73,112),(56,(4,140))),(91,((147,11),(((97,133),((177,27),(144,129))),((64,(125,((90,170),(174,(164,103))))),(35,((115,194),(155,116)))))))))))))))))))))));";
//		String tr1 = "(193,(15,(62,(((99,43),(26,(106,(137,78)))),(((66,(3,(102,(151,165)))),(16,((186,((76,(124,(51,79))),(98,(123,((107,195),(131,(168,104))))))),(((95,121),(9,136)),((6,(67,41)),(81,((45,130),(198,(12,47))))))))),((((153,53),(159,44)),((183,42),((196,22),((19,(40,(14,110))),(156,((173,111),(13,185))))))),((83,((74,148),(146,(109,118)))),(((8,(50,10)),((24,88),((92,38),(132,(75,154))))),((31,((134,(157,82)),((49,72),((188,149),(119,(184,143)))))),(128,(((69,(((29,158),(71,96)),((179,28),((100,87),((126,122),(59,(120,23))))))),((181,((77,30),(152,(70,(85,46))))),((150,(187,(37,61))),((89,182),(161,(135,7)))))),(0,((((113,36),(63,34)),((172,178),((39,65),(68,(166,(60,86)))))),(((145,48),((127,(169,(171,(199,176)))),((32,(18,191)),(180,((142,(105,(117,175))),(((17,93),((162,52),(108,54))),((20,94),((138,(192,160)),(197,(200,141)))))))))),((((58,(114,167)),(57,(139,25))),(2,(((33,1),(190,21)),(5,((80,163),(55,(101,(189,84)))))))),(((73,112),(56,(4,140))),(91,((147,11),(((97,133),((177,27),(144,129))),((64,(125,((90,170),(174,(164,103))))),(35,((115,194),(155,116)))))))))))))))))))))));";
		//(18,(7,2),((3,99),4),((11,(5,9)a,0,22,30)b,6))c;
//		String tr2 = "(9,7,3,4);";
		String tr1 = args[0];
		String tr2 = args[1]; //backbone
		NewickReader nr = new NewickReader(new StringReader(tr1));
		STITree<Double> gt = new STITree<Double>(true);
		nr.readTree(gt);
		
		NewickReader nr2 = new NewickReader(new StringReader(tr2));
		STITree<Double> st = new STITree<Double>(true);
		nr2.readTree(st);
		STINode newNode = new STITree(((STINode<Double>) gt.getNode("18")).toNewick()).getRoot();

		System.err.println(st.toNewick());
		st = treeCompletion(gt,st);
		System.err.println("out: "+st.toNewick());

	}
	
	static STITree addToTree(STITree tree , STINode adoptingNode, STINode toMoveNode){
		
		STINode newNode = null;
		try {
			newNode = new STITree(((STINode<Double>) toMoveNode).toNewick()).getRoot();
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (ParseException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		if(!adoptingNode.isRoot()){
			STINode newinternalnode = adoptingNode.getParent().createChild();
			newinternalnode.adoptChild(adoptingNode);
			newinternalnode.createChild(newNode);
//			if(null != tree.getNode(newNode.getName()))
//				System.err.println("heyyy");
			return tree;
		}
		else{

			STINode newinternalnode = adoptingNode.createChild();
			TNode child = (TNode) adoptingNode.getChildren().iterator().next();
			newinternalnode.adoptChild((TMutableNode) child );
			newinternalnode.createChild(newNode);
			return tree;
			
		}
		
	}

	
static STITree addToTreePolytomy2(STITree tree , STINode adoptingNode, ArrayList<STINode> redChildren){
		
		ArrayList<STINode> newnodes = new ArrayList<STINode>();
		try {
			for(STINode n : redChildren){
				if(!n.isLeaf())
					n.setName("");
				newnodes.add(new STITree(((STINode<Double>)n).toNewick()).getRoot());
		}
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (ParseException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}
		if(adoptingNode.isLeaf()){			
			for(STINode n : newnodes){
				adoptingNode.getParent().createChild(n);
			}
		}
		else{
			for(STINode n : newnodes){
				adoptingNode.createChild(n);
			}			
		}
		return tree;
	}

static STITree addToTreePolytomy(STITree tree , STINode adoptingNode, ArrayList<STINode> redChildren){
		
		ArrayList<STINode> newnodes = new ArrayList<STINode>();
		try {
			for(STINode n : redChildren){
				if(!n.isLeaf())
					n.setName("");
					newnodes.add(new STITree(((STINode<Double>)n).toNewick()).getRoot());
			}
		} catch (IOException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		} catch (ParseException e) {
			// TODO Auto-generated catch block
			e.printStackTrace();
		}

		if(!adoptingNode.isRoot()){
			
			STINode newinternalnode = adoptingNode.getParent().createChild();
			newinternalnode.adoptChild(adoptingNode);
			for(STINode n : newnodes){
				newinternalnode.createChild(n);
			}
			
			return tree;
		}
		else{
				
			STINode newinternalnode = adoptingNode.createChild();
			TNode child = (TNode) adoptingNode.getChildren().iterator().next();
			newinternalnode.adoptChild((TMutableNode) child );
			for(STINode n : newnodes)
				newinternalnode.createChild(n);
			return tree;
			

		}
		
	}
	
	static void nodeColoring(Tree stTree, Tree gtTree) {
		
		
		for (TNode node : stTree.postTraverse()) {
			if(node.isLeaf())
				((STINode) node).setData(new NodeInfo(1));
			else{
				int greenCount = 0;
				for (TNode child:node.getChildren())		
					greenCount += ((NodeInfo) ((STINode) child).getData()).getGrDes();
				((STINode) node).setData(new NodeInfo(greenCount));	
		
			}
		}
			
		Set<String> stLeaves = new HashSet<String>(Arrays.asList(stTree.getLeaves())); 
		
		for (TNode gtNode : gtTree.postTraverse()) {
			if(gtNode.isLeaf()){
				String name = gtNode.getName();
				if (stLeaves.contains(name)){
					((STINode) gtNode).setData(new NodeInfo("G", 1));
				}
				else{
					((STINode) gtNode).setData(new NodeInfo("R", 0));
				}
			}
			else{
				boolean allred = true;
				boolean allgreen = true;
				boolean hasred = false;
				int greenCount = 0;
				for (TNode child: gtNode.getChildren()){
					
					NodeInfo info = (NodeInfo) ((STINode) child).getData();
					String data = info.getColor();
					greenCount += info.getGrDes();
//					String data = (String) ((STINode) child).getData();
					if(data.equals("B") || data.equals("BM") ){
						allred = false;
						allgreen = false;
					}
					else if(data.equals("G")){
						allred = false;
					}
					else if(data.equals("R")){
						allgreen = false;
						hasred = true;
					}
				}
					if(allred) 
						((STINode) gtNode).setData(new NodeInfo("R", greenCount));
					else if(allgreen)
						((STINode) gtNode).setData(new NodeInfo("G", greenCount));
					else if(hasred)
						((STINode) gtNode).setData(new NodeInfo("BM", greenCount));
					else
						((STINode) gtNode).setData(new NodeInfo("B", greenCount));
			}
		}
		
	}
	
	static ArrayList<STITree> treeCompletionRepeat(STITree gTree, STITree sTree, int REPEATS){
		SchieberVishkinLCA lcaLookup = new SchieberVishkinLCA(sTree);
		String[] gtLeaves = gTree.getLeaves();
		String[] stLeaves = sTree.getLeaves();
		List<String> common = new ArrayList<String>(Arrays.asList(gtLeaves));
		common.retainAll(Arrays.asList(stLeaves));
		ArrayList<STITree> results = new ArrayList<STITree>();
		ArrayList<STITree> temps = new ArrayList<STITree>();
		for(int i=0; i< REPEATS; i++){
			temps.add(new STITree(sTree));
		}
		
		ArrayList<Integer> randomRoots = new ArrayList<Integer>();
		for(int i=0;i< REPEATS && i < common.size() ;i++){
			randomRoots.add(GlobalMaps.random.nextInt(common.size()));
		}

		for(int i=0 ; i< REPEATS && i < common.size(); i++){
			String root = common.get(randomRoots.get(i));
			System.err.println("Both constraint tree and gene tree repeat "+i+" are rooted at "+root);
			temps.get(i).rerootTreeAtNode(temps.get(i).getNode(root));
			gTree.rerootTreeAtNode(gTree.getNode(root));
			results.add(treeCompletion(gTree, temps.get(i)));
		}
		return results;
		
		
	}
	static STITree treeCompletion(STITree gTree, STITree sTree){
		nodeColoring(sTree, gTree);
		HashMap<Integer, Integer> LCAMap = createLCAMap(sTree, gTree);
		LCAMap = resolvePolytomies(sTree, gTree, LCAMap);
	  
	        // Create an empty stack and push root to it 
	        Stack<TNode> nodeStack = new Stack<TNode>(); 
	        nodeStack.push(gTree.getRoot());
	        while (nodeStack.empty() == false) { 
	              
	            // Pop the top item from stack and print it 
	            TNode mynode = nodeStack.peek();  
	            NodeInfo info = (NodeInfo) ((STINode) mynode).getData();
				String data = info.getColor();

	            if(data.equals("BM")){
	            	
	            	int childrenCount = mynode.getChildCount();
	            	int redchild = 0;
	            	ArrayList<STINode> redChildren = new ArrayList<STINode>();
	            	ArrayList<STINode> nonRed = new ArrayList<STINode>();
	            	for(TNode child:mynode.getChildren()){
						String childData = (String) ((NodeInfo) ((STINode) child).getData()).getColor();
	            		if(childData.equals("R")){
	            			redChildren.add((STINode) child);
	            			
	            			redchild += 1;
	            		}
	            		else{
	            			nonRed.add((STINode) child);
	            		}
	            	}
	            	int id = LCAMap.get(mynode.getID());
	    	        STINode snode = sTree.getNode(id);
	    	        
	    	        if(nonRed.size()==1 && ((NodeInfo) snode.getData()).getGrDes() == ((NodeInfo) nonRed.get(0).getData()).getGrDes()){
	    	        		sTree = addToTreePolytomy(sTree, snode,  redChildren);
	    	        }
	    	        else{
	    	        		sTree = addToTreePolytomy2(sTree, snode,  redChildren);
	    	        }
	    	        
	    	        //if there is only one non-red child,it means that we definitely don't have this 
	    	        //node in the other tree and we should create a new node
////	    	        if(redchild >= 1 && mynode.getChildCount()-redchild == 1){	
////	    	        	sTree = addToTreePolytomy(sTree, snode,  redChildren);
////	    	        }
//	    	        if(redchild ==1 && mynode.getChildCount() == 2){	            	
//		    	        sTree = addToTree(sTree, snode, (STINode) redChildren.get(0));		    			
//	            	}
//	            	else{
//	            		sTree = addToTreePolytomy(sTree, snode,  redChildren);
//	            	}
	    	        
	    	        //----------------------------
				}
	            nodeStack.pop(); 
	            
	            // Push children of current node to the stack
	            //since at first we added children of the root, we skip it here

	            for(TNode t: mynode.getChildren())
	            		nodeStack.push(t); 

	    }  
	        return sTree;
	}
	
	
	static HashMap<Integer,Integer> resolvePolytomies(Tree stTree, Tree gtTree, HashMap<Integer,Integer> LCAMap){
		for (TNode node : gtTree.postTraverse()) {
			if (node.isLeaf()) {
			}
			else{
            	ArrayList<STINode> nonRed = new ArrayList<STINode>();
            	for(TNode child: node.getChildren()){
					String childData = (String) ((NodeInfo) ((STINode) child).getData()).getColor();
            		if(!childData.equals("R"))     			
            			nonRed.add((STINode) child);
            		
            	}
            	if(nonRed.size() >= 2){
            		STINode snode = (STINode) stTree.getNode(LCAMap.get(node.getID()));
            		//need to add resolution
            		if(snode.getChildCount() > nonRed.size()){
            			boolean sameLCAs = true;
            			ArrayList<STINode> LCAchildren = new ArrayList<STINode>();
            			for(TNode child: nonRed){
            				STINode snodechild = (STINode) stTree.getNode(LCAMap.get(child.getID()));
            				if(!snodechild.getParent().equals(snode)){
            					sameLCAs = false;
            				}
            				else{
            					LCAchildren.add(snodechild);
            				}
            			}
            			if(sameLCAs){
            				STINode newinternalnode = snode.createChild();
            				for(TNode child: LCAchildren){
            					newinternalnode.adoptChild((TMutableNode) child);
            				}
            				LCAMap.put(node.getID(), newinternalnode.getID());
            			}
            			
            		}
            	}
			}
		}
		return LCAMap;
	}
	
	static HashMap<Integer,Integer> createLCAMap(Tree stTree, Tree gtTree) {

		HashMap<Integer, Integer> LCAMap = new HashMap<Integer, Integer>();
		SchieberVishkinLCA lcaLookup = new SchieberVishkinLCA(stTree);

		Stack<TNode> stack = new Stack<TNode>();
		for (TNode gtNode : gtTree.postTraverse()) {
					if (gtNode.isLeaf()) {

						String color = (String) ((NodeInfo) ((STINode) gtNode).getData()).getColor();
						
						if(color.equals("G")){
							TNode t = stTree.getNode(gtNode.getName());
							stack.push(t);
//							LCAMap.put(gtNode.getID(), new ArrayList<Integer>(Arrays.asList(t.getID(),0)));
							LCAMap.put(gtNode.getID(), t.getID());
						}
						if(color.equals("R")){
							stack.push(null);
						}
					} else {
						if(gtNode.getChildCount()==2){
							TNode rightLCA = stack.pop();
							TNode leftLCA = stack.pop();
							// If gene trees are incomplete, we can have this case
							if (rightLCA == null && leftLCA == null) {
								stack.push(null);
								continue;
							}
							else if (rightLCA == null || leftLCA == null) {
								if(rightLCA != null){
									stack.push(rightLCA);
									LCAMap.put(gtNode.getID(), rightLCA.getID());
								}
								if(leftLCA != null){
									stack.push(leftLCA);
									LCAMap.put(gtNode.getID(), leftLCA.getID());
								}
	
								continue;
							}
							
							TNode lca = lcaLookup.getLCA(leftLCA, rightLCA);
							stack.push(lca);
							LCAMap.put(gtNode.getID(), lca.getID());
						}
						else{
							TNode[] children = new TNode[gtNode.getChildCount()];
							boolean allnull = true;
							for(int i =0;i < gtNode.getChildCount();i++){
								children[i] = stack.pop();
								if(children[i] != null)
									allnull = false;
							}
							if (allnull) {
								stack.push(null);
								continue;
							}
							else {
								Set<TNode> notnulls = new HashSet<TNode>();
								for(TNode child: children){
									if(child != null)
										notnulls.add(child);
								}
								TNode lca = lcaLookup.getLCA(notnulls);								
								stack.push(lca);
								LCAMap.put(gtNode.getID(), lca.getID());
								continue;
							}
							
						}

					}
				}
				return LCAMap;
			}
	
	}


	

